spec/consensus/gossip/implementation.md

```diff
@@ -0,0 +1,212 @@
+
+## [REQ-GOSSIP-CONS-SUPERSESSION.2]
+1. Each process only actively participates in one height at a time (proposes/votes);
+1. Each process maintains a maximum number of connections (see configuration parameters).
+1. Broadcast messages are enqueued at most once for each connection on each iteration of the gossip routines.
+1. Unless a message is broadcast multiple times, it will only be sent once on each connection.
+1. GOSSIP keeps track of nodes to which it has sent certain kinds of messages, so resending such messages won't happen even if the message is rebroadcast.
+
+> **TODO** .   
+> Discuss each kind of message and make the previous itemize a proper argument, which will allow showing that $c$ exists, even if high.
+
+
+## P2P-Only nodes
+[REQ-GOSSIP-CONS-SUPERSESSION.2] implies that the messages being broadcast by the process itself and those being forwarded must be limited.
+In Tendermint BFT this is achieved by virtue of only validators broadcasting messages and the set of validators being always limited.
+
+Although only validators broadcast messages, even non-validators (including sentry nodes) must deliver them, because: 
+* only the nodes themselves know if they are validators,
+* non-validators may also need to decide to implement the state machine replication, and,
+* the network is not fully connected and non-validators are used to forward information to the validators.
+
+Non-validators that could support applications above them may be able to inform GOSSIP about superseded messages (for example, upon decisions).
+
+Non-validators that are deployed only to facilitate communication between peers, that is, P2P only nodes that implement GOSSIP but not the CONS, still need to be provided with a supersession operator in order to limit buffering.
+
+
+All nodes currently run Tendermint BFT, but desire to have lightweight, gossip only only, nodes has been expressed, e.g. in [ADR052](#references)
+
+
+## [DEF-BROADCAST]
+1. BMsgs is implemented by the gossip routines encountering certain conditions that will trigger the broadcast of messages.
+1. Broadcast messages are put in outgoing queues for current connections.
+1. If the condition persists, then they are put in the queue again on later iterations, for certain messages.
+1. If the condition persists, then they are put in the queues for new connections in later iterations.
+1. If the condition does not persist, then some superseding messages must have been sent and the superseded one may be dropped (same as remaining in BMsgs but never being actually sent again on the network).
+
+1. Superseding is not explored to drop superseded messages from the outgoing queues.
+
+## [LOCAL_CAST.1]
+For each of the neighbors of the node, looping go-routines continuously evaluate the conditions to send messages to other nodes.
+If a message must be sent, it is enqueued for transmission using TCP and will either be delivered to the destination or the connection will be dropped.
```

Comment:
The condition that a message is either delivered, or the connection is dropped, is not guaranteed, and cannot be assumed.